### ISSUE TITLE
Indirect modification of properties causes no effect

### ISSUE THREAD

**Issue by Ocramius** [goaop/framework#232]
As per today's testing, I found that GO-AOP-PHP proxies are quite weak on magic property access.

Specifically, consider following class:

``` php
<?php

class Foo
{
    private $bar = [];
}

$foo = new Foo;
$foo->bar['baz'] = 'tab';
```

Defining a property access interceptor for this property will likely cause an output like:

```
Notice: Indirect modification of overloaded property Foo::$bar has no effect in /in/lLa8o on line 14 
```

This is because the generated code looks like:

``` php
<?php

class Foo
{
    private $bar = [];

    public function __get($name)
    {
        return $this->bar;
    }
}

$foo = new Foo;
$foo->bar['baz'] = 'tab';
```

The problem is that `__get` is not defined by reference.

I'm taking a ProxyManager generated proxy as a reference implementation (mainly because I tested a lot of aspects of it): https://gist.github.com/Ocramius/55cd7f709c91674add33

What does this mean for the API?
- `& $generated#__get`
- `& Go\Aop\Framework\ClassFieldAccess#__invoke()`
- `& Go\Aop\Framework\ClassFieldAccess#proceed()`

**Comment by Ocramius** [comment]
In general, this needs quite extensive unit testing, specifically on the following operations:
- getting a property
- setting a property
- isset-checking a property 
- unsetting a property
- getting a property array key
- setting a property array key
- isset-checking a property array key
- unsetting a property array key

**Comment by Ocramius** [comment]
Some tests to take as a reference implementation: https://github.com/Ocramius/ProxyManager/blob/7badedb5f8afaaccd496a096daadb19d7bf2b374/tests/ProxyManagerTest/Functional/LazyLoadingGhostFunctionalTest.php#L224-L285

**Comment by lisachenko** [comment]
Thanks, @Ocramius, this is known issue #54, actually it isn't fully baked now like in your ProxyManager. I have never seen before the real use-cases for intercepting property access, because it's very rare case ) Your check is real first one that requires an interceptions of properties. 

My proxy is very limited indeed (no `__isset`, `__unset`, no reference passing), so a lot of thing to do and test.

So can I just your library to generate a proxy with my method body? This can be slower, but will more reliable and extendable.

**Comment by Ocramius** [comment]
As it currently stands, the problem is that `ProxyManager` uses runtime reflection, which is not compatible with how GO-AOP-PHP works (runtime code generation).

@asgrim is working on https://github.com/Roave/BetterReflection, which is a "cold" reflector (doesn't autoload classes, and would fit this particular use-case).

Another approach (for now) is using a separate process for the codegen.

**Comment by lisachenko** [comment]
Implemented in #281, property interceptors work by reference now.
